### PR TITLE
Mailbox rent

### PR DESCRIPTION
**Pull request by breathx**
Part of #646.
Part of #899.
Part of #1130.

Changes:
- Found and resolved bug on resetting storage: added `reset` to gas handler interface.
- Resolved clippy issue in property tests.
- Value transferring moved from messenger callbacks to gear pallet.
- Added `IterableMap` implementation for `Mailbox`.
- Added zero-copy ability to convert message into message with string payload.
- Internal module refactored in order of simplification and well documented.
- Added proper mailbox rent calculation and charging: task processing, and so on.
- Everywhere used normal `MailboxThreshold`.
- Added tests for all cases.

**Release notes**: _now we properly charge rent in mailbox._

### CODE REVIEW

File: common/src/gas_provider/property_tests/utils.rs [hunk 1]
```diff
@@ -28,9 +28,8 @@ pub(super) trait RingGet<T> {
 
 impl<T> RingGet<T> for Vec<T> {
     fn ring_get(&self, index: usize) -> Option<&T> {
-        let is_not_empty = !self.is_empty();
-        is_not_empty
-            .then(|| index % self.len())
+        (!self.is_empty())
+            .then_some(index % self.len().max(1))
```
Comment (SabaunT) [inline common/src/gas_provider/property_tests/utils.rs:32]: smart :clap:

**Review by shamilsan — COMMENTED** [review]
Extremely hard to review this huge PR. So just minor notes.

File: core/src/message/common.rs [hunk 2]
```diff
@@ -133,6 +134,18 @@ impl Message {
     pub fn exit_code(&self) -> Option<ExitCode> {
         self.reply.map(|v| v.exit_code())
     }
+
+    /// Consumes self in order to create new `StoredMessage`, which payload
+    /// contains string representation of initial bytes,
+    /// decoded into given type.
+    pub fn with_string_payload<D: Decode + ToString>(self) -> Result<Self, Self> {
```
Comment (grishasobol) [inline core/src/message/common.rs:141]: It's not used anywhere ?

File: runtime/src/lib.rs [hunk 3]
```diff
@@ -351,7 +351,7 @@ impl pallet_gear::Config for Runtime {
     type OutgoingLimit = ConstU32<1024>;
     type DebugInfo = DebugInfo;
     type CodeStorage = GearProgram;
-    type MailboxThreshold = ConstU64<0>;
+    type MailboxThreshold = ConstU64<3000>;
```
Comment (grishasobol) [inline runtime/src/lib.rs:354]: How did you choose this value?

File: runtime/src/lib.rs [hunk 4]
```diff
@@ -370,14 +370,14 @@ impl pallet_gear_scheduler::Config for Runtime {
     type BlockLimiter = GearGas;
     type ReserveThreshold = ConstU32<1>;
     type WaitlistCost = ConstU64<100>;
+    type MailboxCost = ConstU64<100>;
```
Comment (grishasobol) [inline runtime/src/lib.rs:373]: How did you choose this value?

**Review by grishasobol — APPROVED** [review]

File: pallets/gear/src/internal.rs [hunk 5]
```diff
@@ -341,12 +377,178 @@ where
...
+    }
+
+    /// Charges and deposits event for already taken from mailbox message.
+    ///
+    /// Note: message auto-consumes, if reason is claim or reply.
+    pub(crate) fn read_message_requirements(
+        (mailboxed, hold_interval): (StoredMessage, Interval<BlockNumberFor<T>>),
+        reason: UserMessageReadReason,
+    ) -> StoredMessage {
+        // Local import for code beautification.
```
Comment (gshep) [inline pallets/gear/src/internal.rs:411]: I guess this is unnecessary comment

File: pallets/gear/src/manager/journal.rs [hunk 6]
```diff
@@ -188,24 +183,24 @@ where
...
-        );
+            log::debug!(
+                "Sending message {:?} from {:?} with gas limit {:?}",
+                dispatch.message(),
+                message_id,
+                gas_limit,
+            );
+
+            if dispatch.value() != 0 {
+                <T as Config>::Currency::reserve(
```
Comment (gshep) [inline pallets/gear/src/manager/journal.rs:198]: `CurrencyOf`

File: pallets/gear/src/tests.rs [hunk 7]
```diff
@@ -108,6 +113,305 @@ fn unstoppable_block_execution_works() {
...
+
+        run_to_next_block(None);
+
+        assert!(Gear::is_initialized(sender));
+
+        // FOR BOTH CASES VALUE MOVES TO DESTINATION USER.
+        let cases = [
+            // CASE WHEN GASFUL MESSAGE REMOVED DUE TO CLAIM.
+            (TestData::gasful(20_000, 1_000), 10),
+            // CASE WHEN GASLESS MESSAGE REMOVED DUE TO CLAIM.
```
Comment (gshep) [inline pallets/gear/src/tests.rs:241]: why caps-locked?

**Review by gshep — APPROVED** [review]
Good job! :gear:

File: core/src/message/common.rs (hunk 2 above)
Comment (breathx) [reply core/src/message/common.rs:141]: Here it is in test tools:
https://github.com/gear-tech/gear/blob/4927c9c2c355c71394dfd568371e35ae2270de90/gtest/src/manager.rs#L674-L679
And on-chain:
https://github.com/gear-tech/gear/blob/25b791809af67e64b761ca7220bda47a61ffb73c/pallets/gear/src/internal.rs#L491-L496

File: runtime/src/lib.rs (hunk 3 above)
Comment (breathx) [reply runtime/src/lib.rs:359]: Random for now.
This needs to be discussed/benchmarked somehow before release

File: runtime/src/lib.rs (hunk 4 above)
Comment (breathx) [reply runtime/src/lib.rs:378]: Same as mailbox threshold. May be changed any time without need to update tests etc.

File: core/src/message/common.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (breathx) [reply core/src/message/common.rs:141]: Discussed. I see no reason to add crate dependency for only one function

File: examples/binaries/value-sender/src/lib.rs [hunk 8]
```diff
@@ -0,0 +1,104 @@
...
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see <https://www.gnu.org/licenses/>.
+
+#![no_std]
+
+use codec::{Decode, Encode};
+use gstd::ActorId;
```
[the remark this answers is not in the document]
Comment (breathx) [reply examples/binaries/value-sender/src/lib.rs:22]: Discussed. In-place one demo for test for me is more readable.

File: pallets/gear/src/internal.rs [hunk 9]
```diff
@@ -19,67 +19,186 @@
...
+
+// `unused` allowed because some fns may be used in future, but clippy
+// doesn't allow this due to `pub(crate)` visibility.
+#[allow(unused)]
+impl<T: Config> HoldBound<T> {
+    /// Creates cost builder for hold bound.
+    pub fn by(cost: SchedulingCostOf<T>) -> HoldBoundCost<T> {
+        assert!(!cost.is_zero());
+        HoldBoundCost(cost)
+    }
```
[the remark this answers is not in the document]
Comment (breathx) [reply pallets/gear/src/internal.rs:122]: I don't really love builder pattern as it used usual, but I like it small representation:
I build `HoldBound` and to make code readable, I include such builder in it
`HoldBound::by(COST).anything_need_further()`

File: pallets/gear/src/internal.rs [hunk 10]
```diff
@@ -19,67 +19,186 @@
...
+        self.at(expected)
+    }
+
+    /// Creates maximal available bound for given gas limit.
+    pub fn maximum_for(self, gas: GasBalanceOf<T>) -> HoldBound<T> {
+        let deadline_duration = gas
+            .saturating_div(self.0.max(One::one()))
+            .saturated_into::<BlockNumberFor<T>>();
+
+        let deadline = SystemPallet::<T>::block_number().saturating_add(deadline_duration);
```
[the remark this answers is not in the document]
Comment (breathx) [reply pallets/gear/src/internal.rs:80]: Discussed. It's matched in `deadline` constructor

**Pull request merged**